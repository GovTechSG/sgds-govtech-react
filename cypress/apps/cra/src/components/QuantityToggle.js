import { useState } from "react";
import { Row } from '@govtechsg/sgds-react/Row'
import { Col } from '@govtechsg/sgds-react/Col'
import { Card } from '@govtechsg/sgds-react/Card'
import { CloseButton } from '@govtechsg/sgds-react/CloseButton'
import { QuantityToggle } from '@govtechsg/sgds-react/QuantityToggle'
const QtyToggleCom = () => {
    const [count, setCount] = useState(0);
    const [disappear, setDisappear] = useState(false);
    const handleDismiss = () => {
      setDisappear(!disappear);
    };
    if (!disappear)
      return (
        <Row>
          <Col xs lg="5">
            <Card variant="card-action-quantity-toggle">
              <Card.Body>
                <div>
                  <Card.Title as="h4">
                    Basket <i className="bi bi-cart"></i>
                  </Card.Title>
                  <CloseButton onClick={handleDismiss} />
                </div>
                <div>
                  <Card.Unit>
                    <Card.Subtitle as="h6" className="text-muted">
                      Item Brand
                    </Card.Subtitle>
                    <Card.Subtitle as="h6">Apple</Card.Subtitle>
                  </Card.Unit>
                  <QuantityToggle count={count} setCount={setCount} />
                </div>
              </Card.Body>
            </Card>
          </Col>
        </Row>
      );
    else return null;
  };
  export default QtyToggleCom