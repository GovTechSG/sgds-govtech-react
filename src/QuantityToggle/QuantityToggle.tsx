import * as React from 'react';
import Button from '../Button/Button';
import InputGroup from '../InputGroup/InputGroup';
import FormControl from '../Form/FormControl';
import { BsPrefixRefForwardingComponent } from '../utils/helpers';
import { ButtonVariant } from '../utils/types';
import PropTypes from 'prop-types';
import { FormLabel } from '../Form';

export interface QuantityToggleProps {
  /** Controls the incremental / decremental value */
  step?: number;
  /** Controls the size of the QuantityToggle */
  size?: 'sm' | 'lg';
  /**Keep tracks of input value. This parameter should be passed down from parent with `setCount` */
  count: number;
  /**Updates the count value in Parent component. This parameter should be passed down from parent with `count` */
  setCount: React.Dispatch<React.SetStateAction<number>>;
  /** Disables the buttons and input of `QuantityToggle` */
  disabled?: boolean;
  /**Controls the color of the buttons */
  variant?: ButtonVariant;
}

const defaultProps: Partial<QuantityToggleProps> = {
  size: 'sm',
  step: 1,
};
const propTypes = {
  step: PropTypes.number,
  size: PropTypes.oneOf(['sm', 'lg']),
  count: PropTypes.number.isRequired,
  setCount: PropTypes.func.isRequired,
  disabled: PropTypes.bool,
  variant: PropTypes.oneOf<ButtonVariant>([
    'primary',
    'secondary',
    'success',
    'danger',
    'warning',
    'info',
    'dark',
    'light',
    'link',
    'outline-primary',
    'outline-secondary',
    'outline-success',
    'outline-danger',
    'outline-warning',
    'outline-info',
    'outline-dark',
    'outline-light',
  ]),
};

export const QuantityToggle: BsPrefixRefForwardingComponent<
  'input',
  QuantityToggleProps
> = React.forwardRef<HTMLInputElement, QuantityToggleProps>(
  ({ size, step = 1, disabled, variant, count, setCount, ...props }, ref) => {
    const buttonProps = { disabled, variant };
    const onPlus = () => {
      setCount(count + step);
    };
    const onMinus = () => {
      if (count < 1) setCount(0);
      else setCount(count - step);
    };
    React.useEffect(() => {
      if (count < 0) setCount(0);
    }, []);
    return (
      <>
        <FormLabel className="visually-hidden">quantity-toggle</FormLabel>
        <InputGroup size={size} variant="quantity-toggle">
          <Button onClick={onMinus} {...buttonProps} aria-label="decrement-button">
            <i className="bi bi-dash"></i>
          </Button>
          <FormControl
            {...props}
            disabled={disabled}
            ref={ref}
            type="number"
            className="text-center"
            value={count}
            name="quantity"
            onChange={(e) => {
              setCount(parseInt(e.target.value));
            }}
            min={0}
          />
          <div
            id="quantitytoggle-announcer"
            role="region"
            aria-live="assertive"
            className="visually-hidden">
            {count}
          </div>
          <Button onClick={onPlus} {...buttonProps} aria-label="increment-button">
            <i className="bi bi-plus"></i>
          </Button>
        </InputGroup>
      </>
    );
  }
);

QuantityToggle.displayName = 'QuantityToggle';
QuantityToggle.defaultProps = defaultProps;
QuantityToggle.propTypes = propTypes;

export default QuantityToggle;
