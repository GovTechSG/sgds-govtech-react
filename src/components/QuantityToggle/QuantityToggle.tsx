import * as React from 'react';
import { useState } from 'react';
import Button, { ButtonProps } from '../Button/Button';
import InputGroup, { InputGroupProps } from '../InputGroup/InputGroup';
import FormControl, { FormControlProps } from '../Form/FormControl';

type QuantityToggleType = Omit<InputGroupProps, 'hasValidation'> & Omit<ButtonProps, 'size'> & Omit<FormControlProps, 'isValid' | 'isInvalid' |'searchIcon'>;

export interface QuantityToggleProps extends QuantityToggleType {
    min?: number;
    step?: number;
    size?: 'sm' | 'lg'
}

const defaultProps: Partial<QuantityToggleProps> = {
    size: 'sm',
    min: 0,
    step: 1
}

const QuantityToggle: React.FC<QuantityToggleProps> = ({size, min = 0 , step = 1,  active, variant, ...props}) => {
  const [count, setCount] = useState(min);
  const buttonProps = {active, variant}
  const onPlus = () => {
    setCount(count + step);
  };
  const onMinus = () => {
    if (count < 1) setCount(0);
    else setCount(count - step);
  };
  return (
    <InputGroup size={size}>
      <Button onClick={onMinus} {...buttonProps}>-</Button>
      <FormControl
       {...props}
        type="number"
        value={count}
        onChange={(e) => {
          setCount(parseInt(e.target.value));
        }}
      />
      <Button onClick={onPlus}>+</Button>
    </InputGroup>
  );
};

QuantityToggle.displayName = 'QuantityToggle'
QuantityToggle.defaultProps = defaultProps

export default QuantityToggle