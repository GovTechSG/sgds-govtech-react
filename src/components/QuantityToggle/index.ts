export {default as QuantityToggle} from './QuantityToggle'
export type { QuantityToggleProps } from './QuantityToggle'