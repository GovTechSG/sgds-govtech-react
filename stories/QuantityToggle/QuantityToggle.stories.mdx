import { ArgsTable, Canvas, Meta, Story } from '@storybook/addon-docs';
import { useState, useEffect } from 'react';
import { QuantityToggle, Col } from '../../src/components/index';
import { HIDDEN_COMMON_PROPS, VARIANT } from '../helpers';

<Meta
  title="MDX/QuantityToggle"
  component={QuantityToggle}
  argTypes={{
    ...HIDDEN_COMMON_PROPS,
    step: {
      description: 'Controls the incremental / decremental value',
    },
    size: { description: 'Controls the size of the QuantityToggle' },
    count: {
      description:
        'Keep tracks of input value. This parameter should be passed down from parent with `setCount`',
    },
    setCount: {
      description:
        'Updates the count value in Parent component. This parameter should be passed down from parent with `count`',
    },
    disabled: {
      description: 'Disables the buttons and input of `QuantityToggle`',
    },
    variant: {
      description: 'Controls the color of the buttons',
      control: { type: 'select' },
      options: VARIANT,
    },
  }}
/>

# Quantity Toggle

`<QuantityToggle />` allow increments and decrement of value reflected on an `input`. 

It is a custom component built from `InputGroup`, `Button`, `FormControl`

The default `QuantityToggle` requires `count` and `setCount` to be passed in as props.
The default step value is 1 which controls the increment / decrement value on every button click.
The `input` is of number type. 

<Canvas>
  <Story name="Default">
    {(args) => {
      const [count, setCount] = useState(0);
      return (
          <Col sm={2}>
          <QuantityToggle {...args} count={count} setCount={setCount} />
          </Col>
      );
    }}
  </Story>
</Canvas>

## API

<ArgsTable story="Default" />
