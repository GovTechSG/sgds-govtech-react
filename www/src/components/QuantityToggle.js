import { useState } from "react";
import { Row } from '@govtechsg/sgds-govtech-react/Row'
import { Col } from '@govtechsg/sgds-govtech-react/Col'
import { Card } from '@govtechsg/sgds-govtech-react/Card'
import { CloseButton } from '@govtechsg/sgds-govtech-react/CloseButton'
import { QuantityToggle } from '@govtechsg/sgds-govtech-react/QuantityToggle'
const QtyToggleCom = () => {
    const [count, setCount] = useState(0);
    const [disappear, setDisappear] = useState(false);
    const handleDismiss = () => {
      setDisappear(!disappear);
    };
    if (!disappear)
      return (
        <Row>
          <Col xs lg="5">
            <Card variant="card-action-quantity-toggle">
              <Card.Body>
                <div>
                  <Card.Title as="h4">
                    Basket <i className="bi bi-cart"></i>
                  </Card.Title>
                  <CloseButton onClick={handleDismiss} />
                </div>
                <div>
                  <Card.Unit>
                    <Card.Subtitle as="h6" className="text-muted">
                      Item Brand
                    </Card.Subtitle>
                    <Card.Subtitle as="h6">Apple</Card.Subtitle>
                  </Card.Unit>
                  <QuantityToggle count={count} setCount={setCount} />
                </div>
              </Card.Body>
            </Card>
          </Col>
        </Row>
      );
    else return null;
  };
  export default QtyToggleCom