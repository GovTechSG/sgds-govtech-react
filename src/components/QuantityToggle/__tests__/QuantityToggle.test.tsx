import * as React from 'react';
import QuantityToggle from '../QuantityToggle'
import {  fireEvent, render } from '@testing-library/react';

describe('<QuantityToggle />', () => {
    it('has default structure', () => {
        const mockSetCount = jest.fn()
        const {asFragment, getByText, container } = render(<QuantityToggle count={0} setCount={mockSetCount}/>)

        expect(asFragment()).toMatchSnapshot()
        expect(container.querySelector('div')).toBeDefined()
        const $InputGroup = container.querySelector('div')
        expect($InputGroup?.classList).toContain('input-group')
        expect($InputGroup?.children.length).toEqual(3)
        expect($InputGroup?.children[0].tagName).toEqual('BUTTON')
        expect($InputGroup?.children[1].tagName).toEqual('INPUT')
        expect($InputGroup?.children[2].tagName).toEqual('BUTTON')

        const $input = $InputGroup?.children[1]
        expect($input).toHaveAttribute('type', 'number')
        expect($input).toHaveAttribute('name', 'quantity')
        expect($input).toHaveAttribute('value', "0") 
        expect(getByText('-').tagName).toEqual('BUTTON')
        expect(getByText('+').tagName).toEqual('BUTTON')
    })
    it('onClick of buttons triggers setCount mock', () => {
        const mockSetCount = jest.fn()
        const { getByText } = render(<QuantityToggle count={0} setCount={mockSetCount}/>)
        fireEvent.click(getByText('+'))
        expect(mockSetCount).toHaveBeenCalledTimes(1)
        fireEvent.click(getByText('-'))
        expect(mockSetCount).toHaveBeenCalledTimes(2)

    })
})