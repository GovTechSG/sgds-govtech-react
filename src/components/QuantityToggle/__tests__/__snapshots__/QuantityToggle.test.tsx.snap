// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<QuantityToggle /> has default structure 1`] = `
<DocumentFragment>
  <div
    class="input-group input-group-sm"
  >
    <button
      class="btn btn-primary sgds"
      type="button"
    >
      -
    </button>
    <input
      class="form-control"
      name="quantity"
      type="number"
      value="0"
    />
    <button
      class="btn btn-primary sgds"
      type="button"
    >
      +
    </button>
  </div>
</DocumentFragment>
`;
